src/Parser/DestructionWarlock/Modules/SoulShards/SoulShardEvents.js

```diff
@@ -0,0 +1,254 @@
+  [SPELLS.SUMMON_DOOMGUARD_TALENTED.id]: SUMMON_COST,
+  [SPELLS.GRIMOIRE_VOIDWALKER.id]: SUMMON_COST,
+  [SPELLS.GRIMOIRE_SUCCUBUS.id]: SUMMON_COST,
+  [SPELLS.GRIMOIRE_FELHUNTER.id]: SUMMON_COST,
+  [SPELLS.SUMMON_VOIDWALKER.id]: SUMMON_COST,
+  [SPELLS.SUMMON_SUCCUBUS.id]: SUMMON_COST,
+  [SPELLS.SUMMON_FELHUNTER.id]: SUMMON_COST,
+};
+
+const FRAGMENT_GENERATING_ENERGIZE_ABILITIES = [
+  SPELLS.SOUL_CONDUIT_SHARD_GEN.id,
+  SPELLS.SOULSNATCHER_FRAGMENT_GEN.id,
+  SPELLS.FERETORY_OF_SOULS_FRAGMENT_GEN.id,
+];
+
+class SoulShardEvents extends Module {
+  static dependencies = {
+    enemies: Enemies,
+    combatants: Combatants,
+  };
+
+  hasT20_2p = false;
+  currentFragments = 0;
+
+  on_initialized() {
+    if (!this.owner.error) {
+      this.hasT20_2p = this.combatants.selected.hasBuff(SPELLS.WARLOCK_DESTRO_T20_2P_BONUS.id);
+      this.currentFragments = 30; //on the start of the fight we should have 3 soul shards (30 fragments) by default
+      debug && console.log("start fragments " + this.currentFragments);
+    }
+  }
+
+  on_byPlayer_energize(event) {
+    if (event.resourceChangeType !== RESOURCE_TYPES.SOUL_SHARDS) {
+      return;
+    }
+    const spellId = event.ability.guid;
+    if (FRAGMENT_GENERATING_ENERGIZE_ABILITIES.indexOf(spellId) === -1) {
+      return;
+    }
+
+    const shardEvent = {
+      timestamp: event.timestamp,
+      type: 'soulshardfragment_gained',
+      ability: {
+        guid: spellId,
+        name: SPELLS[spellId].name,
+      },
+    };
+    //purposefully missing targetID and targetInstance because both source and target on energize is the player
+
+    const rawGain = event.resourceChange * FRAGMENTS_PER_SHARD;
+    let gain = 0;
+    let waste = 0;
+    if (this.currentFragments + rawGain > MAX_FRAGMENTS) {
+      gain = MAX_FRAGMENTS - this.currentFragments;
+      waste = this.currentFragments + rawGain - MAX_FRAGMENTS;
+    }
+    else {
+      gain = rawGain;
+    }
+
+    this.currentFragments += gain;
+
+    shardEvent.amount = gain;
+    shardEvent.waste = waste;
+    shardEvent.currentFragments = this.currentFragments;
+
+    debug && console.log('++ ' + shardEvent.amount + '(w: ' + shardEvent.waste + ') = ' + shardEvent.currentFragments + ', ' + shardEvent.ability.name + ', orig: ', event);
+    this.owner.triggerEvent('soulshardfragment_gained', shardEvent);
+  }
+
+  //should also work with Havoc and talent Fire and Brimstone
+  on_byPlayer_damage(event) {
+    const spellId = event.ability.guid;
+    if (FRAGMENT_GENERATING_ABILITIES.indexOf(spellId) === -1) {
+      return;
+    }
+
+    const shardEvent = {
+      timestamp: event.timestamp,
+      type: 'soulshardfragment_gained',
+      ability: {
+        guid: spellId,
+        name: SPELLS[spellId].name,
+      },
+      targetID: event.targetID,
+      targetInstance: event.targetInstance,
+    };
+    let rawGain;
+    let gain = 0;
+    let waste = 0;
+
+    switch (spellId) {
```

Comment:
Purely opinion here, but I think object literals are easier to use and read for this,

```javascript
const SHARD_GENERATION = {
  [SPELLS.CONFLAGRATE.id]: 5,
  ...
};

const rawGain = SHARD_GENERATION[spellid];
```

Not only are you not overriding the value of `rawGain`, but you also have an easy to maintain dictionary of shard generation values. This isn't a blocker.